nav2_costmap_2d/test/unit/keepout_filter_test.cpp

```diff
@@ -0,0 +1,363 @@
+  void testKeepoutFilter(unsigned char free_value, unsigned char keepout_value);
+
+  std::shared_ptr<nav2_costmap_2d::KeepoutFilter> keepout_filter_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> master_grid_;
+
+  std::vector<Point> keepout_points_;
+
+private:
+  void verifyMasterGrid(unsigned char free_value, unsigned char keepout_value);
+
+  nav2_util::LifecycleNode::SharedPtr node_;
+
+  std::shared_ptr<nav_msgs::msg::OccupancyGrid> mask_;
+
+  std::shared_ptr<InfoPublisher> info_publisher_;
+  std::shared_ptr<MaskPublisher> mask_publisher_;
+};
+
+void TestNode::createMaps(unsigned char master_value, int8_t mask_value)
+{
+  // Make map and mask overlapping as follows:
+  //
+  //        map     (10,10)
+  //         *---------*
+  //   mask  |  (5,5)  |
+  //    *----+----*    |
+  //    *////|////*    |
+  //    *////*----*----*
+  //    */////////*
+  //    *---------*
+  // (-5,-5)
+
+  const unsigned int width = 10;
+  const unsigned int height = 10;
+  const double resolution = 1.0;
+
+  // Create master_grid_
+  master_grid_ = std::make_shared<nav2_costmap_2d::Costmap2D>(
+    width, height, resolution, 0.0, 0.0, master_value);
+
+  // Create mask_
+  mask_ = std::make_shared<nav_msgs::msg::OccupancyGrid>();
+  mask_->info.resolution = resolution;
+  mask_->info.width = width;
+  mask_->info.height = height;
+  mask_->info.origin.position.x = -5.0;
+  mask_->info.origin.position.y = -5.0;
+  mask_->info.origin.position.z = 0.0;
+  mask_->info.origin.orientation.x = 0.0;
+  mask_->info.origin.orientation.y = 0.0;
+  mask_->info.origin.orientation.z = 0.0;
+  mask_->info.origin.orientation.w = 1.0;
+  mask_->data.resize(width * height, mask_value);
+}
+
+void TestNode::publishMaps()
+{
+  info_publisher_ = std::make_shared<InfoPublisher>();
+  while (!info_publisher_->isPublished()) {
+    rclcpp::spin_some(info_publisher_);
+    std::this_thread::sleep_for(100ms);
+  }
+
+  mask_publisher_ = std::make_shared<MaskPublisher>(*mask_);
+  while (!mask_publisher_->isPublished()) {
+    rclcpp::spin_some(mask_publisher_);
+    std::this_thread::sleep_for(100ms);
+  }
+}
+
+void TestNode::reset()
+{
+  mask_.reset();
+  master_grid_.reset();
+  info_publisher_.reset();
+  mask_publisher_.reset();
+  keepout_filter_.reset();
+  node_.reset();
+  keepout_points_.clear();
+}
+
+void TestNode::verifyMasterGrid(unsigned char free_value, unsigned char keepout_value)
+{
+  unsigned int x, y;
+  bool is_checked;
+
+  for (y = 0; y < master_grid_->getSizeInCellsY(); y++) {
+    for (x = 0; x < master_grid_->getSizeInCellsX(); x++) {
+      is_checked = false;
+      for (std::vector<Point>::iterator it = keepout_points_.begin();
+        it != keepout_points_.end(); it++)
+      {
+        if (x == it->x && y == it->y) {
+          EXPECT_EQ(master_grid_->getCost(x, y), keepout_value);
+          is_checked = true;
+          break;
+        }
+      }
+      if (!is_checked) {
+        EXPECT_EQ(master_grid_->getCost(x, y), free_value);
```

Comment:
What's this check for?

Reply:
This check is intended to verify that all other than `keepout_points_` points on `master_grid` are not "stained" by a filter. In some cases, e.g. when bounds are not calculated correctly, `KeepoutFilter` might mark superfluous points on `master_grid` as `keepout_value`. In other words, this check is counted on the fact that the points from `keepout_points_` are set and only them having `keepout_value`. All other points on `master_grid` should remain to be untouched.

Reply:
ah ok, makes sense.